### DISCUSSION TITLE
Expand `where T : new()` constructor constraint to allow known parameter types

### DISCUSSION THREAD

**Discussion by thomhurst** [dotnet/csharplang#8542] [Language Ideas] [answered]
I don't see any reason this couldn't be done. The compiler will be able to look at the public constructors of a type and know if those are available.

This can be useful when wanting to instantiate generic unknown types but with well-known parameters.

A good example of this could be custom exceptions that follow the standard exception constructor patterns.

Syntax:

`where TType : new(TParameterTypes..)` where TParameterTypes can be as many types as you define, comma separated.

E.g.

`where TType : new(TParameterType1, TParameterType2, TParameterType3)` 

Use case example:

```csharp
public class GenericFailureHandler<TExceptionType>
    where TExceptionType : Exception
    where TExceptionType : new(string)
    where TExceptionType : new(string, Exception)
{
    public void Handle(Exception parentException)
    {
        ...
        throw new TExceptionType("My reason", parentException);
    }
}
```

If `TExceptionType` didn't have a constructor taking a `string`, and a constructor taking a `string, Exception`, then the compiler would give you an error, much like it does with `where T : new()` when there isn't a parameterless constructor.

This can avoid the `Activator` classes, and eliminate their need for reflection in cases where the parameters are known.

This also means it would help with Native AOT if the compiler knows about the constructor usage at compile time.

**Comment by colejohnson66** [comment]
Logically, yes, this makes sense. But the IL does not support it. Generic constrains *must* be one or more of the following:

* a covariant type
* a contravariant type
* `class`
* `valuetype` (`struct` in C#)
* `.ctor` (`new()` in C#)

See II.10.1.7 in ECMA-335 (6th edition)

![IMG_1385](https://github.com/user-attachments/assets/dae2f2ca-6ca9-42bf-9e4d-22fa74b8f1a2)

**Comment by huoyaoyuan** [comment, marked as answer]
#769, #1330, #1574, #1626, #1633, #4532
